### ISSUE TITLE
Rust and C versions of KangarooTwelve and MarsupilamiFourteen 

### ISSUE THREAD

**Issue by DonaldTsang** [XKCP/XKCP#54]

**Comment by newpavlov** [comment]
I've added MarsupilamiFourteen to the list of missing hash functions.

**Comment by DonaldTsang** [comment]
@newpavlov I meant the secondary reference for XKCP

**Comment by newpavlov** [comment]
Sorry, I don't quite understand you. k12 implementation is not yet implemented in RustCrypto, and if we'll get MarsupilamiFourteen I think it will be a separate crate.

**Comment by DonaldTsang** [comment]
I meant that K12 and MarsupilamiFourteen should be implemented in XKCP (as it is owned by the official Keccak Team)

**Comment by newpavlov** [comment]
Well, I am not sure why you need my permission for it. But to improve discoverability and to be able to evolve API coherently I plan to keep our own implementation under RustCrypto organization. If XKCP code will be better, we will use the relevant code bits in our crates, and you can do vice versa.

**Comment by DonaldTsang** [comment]
@newpavlov whose code is this then? https://github.com/XKCP/XKCP/blob/master/Standalone/KangarooTwelve/Rust/k12.rs
If it is copied from your repository then you should be able to understand your code, right?

**Comment by dhardy** [comment]
@DonaldTsang yes, I believe that is my code, but with some code copied from the two links at the top of that file. I recall there were some odd things about the Python code I used as reference — not sure if these were bugs or what (see TODO comments).

But sure, feel free to adapt it for whatever Marsupial is now involved...

**Comment by DonaldTsang** [comment]
@dhardy in order to edit that first we need to know where are the 2 missing constants to move from K12 to M14. Then we have to add extra steps to allow for variant-round scrambling (up to 24)

**Comment by dhardy** [comment]
@DonaldTsang [this](https://github.com/dhardy/hash-bench/blob/master/src/k12.rs) is the version I wrote with history, which may or may not help. I don't have the time to get into this now so it's up to you.

**Comment by gvanas** [comment]
KT256 superseeds MarsupilamiFourteen.
